chainstate/src/detail/chainstateref/mod.rs

```diff
@@ -428,36 +427,54 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
         Ok(())
     }
 
-    pub fn check_block_header(
-        &self,
-        header: &SignedBlockHeader,
-        strict_pos_consensus_check: bool,
-    ) -> Result<(), CheckBlockError> {
+    pub fn check_block_header(&self, header: &SignedBlockHeader) -> Result<(), CheckBlockError> {
         self.check_header_size(header).log_err()?;
         self.enforce_checkpoints(header).log_err()?;
 
-        // TODO(Gosha):
-        // using utxo set like this is incorrect, because it represents the state of the mainchain, so it won't
-        // work when checking blocks from branches. See mintlayer/mintlayer-core/issues/752 for details
         let utxos_db = UtxosDB::new(&self.db_tx);
         let pos_db = PoSAccountingDB::<_, TipStorageTag>::new(&self.db_tx);
+
+        let best_block_id = self.get_best_block_id()?;
+        let (utxos_cache, pos_delta, epoch_data_cache) = if best_block_id == *header.prev_block_id()
```

Comment:
Keep in mind that this reorging is only necessary for select types of consensus. Perhaps we should wrap this whole thing with a boolean derived from ConsensusData, where it's only true for PoS. We don't need to reorg for PoW.

Reply:
added check for pos